Add RPC chapter (#30)
tikv/deep-dive-tikv: src/rpc/introduction.md

# Remote Procedure Calls (RPC)

> Communication between services occurs over remote procedure calls. RPCs happen all the time in distributed systesm. You browser had to make at least one RPC to this website to obtain this webpage.

Reviewer: ```suggestion
Communication between processes can use remote procedure calls. RPCs happen all the time in distributed systems. You browser had to make at least one RPC to this website to obtain this webpage.
```

I'm not sure if HTTP is a kind of RPC, I don't think I would classify it as such. For me, RPC implies a procedure-like interface, rather than a networking protocol style interface.

TiKV, as a distributed system involving a number of nodes, uses RPCs to communicate between each node, as well as the placement driver and clients.

> As it turns out, exposing functionality as a remote interface isn't a trivial. Networks are unreliable, systems are diverse and evolving, a huge variety of languages and formats exist, and even things like encoding are hard!

Reviewer: ```suggestion
As it turns out, exposing functionality as a remote interface isn't trivial. Networks are unreliable, systems are diverse and evolving, a huge variety of languages and formats exist, and even things like encoding are hard!
```

I would add something like: trying to make all these issues transparent (i.e., treating RPCs like local calls) has proven to be impossible, but making all the networking stuff explicit makes the communication code verbose and error-prone

## On the shoulders of giants

> Over the past decades the field of computing has largely settled on a few common standards.

Reviewer: Unfortunately, I think this is overstating it - there are a lot of ways to do IPC/network communication and it doesn't seem like we've settled on much above the network protocols

### Network Protocols

> The vast majority of services work over the HTTP or HTTP/2 network protocols. This solves problems such as:

Reviewer: I'm not sure exactly what you mean by services (should be clarified) but there are huge amounts of networking 'stuff' which don't use HTTP - email, FTP, UDP, etc.

# Remote Procedure Calls (RPC)

Communication between services occurs over remote procedure calls. RPCs happen
all the time in distributed systesm. You browser had to make at least one RPC to
this website to obtain this webpage.

TiKV, as a distributed system involving a number of nodes, uses RPCs to
communicate between each node, as well as the placement driver and clients.

As it turns out, exposing functionality as a remote interface isn't a trivial.
Networks are unreliable, systems are diverse and evolving, a huge variety of
languages and formats exist, and even things like encoding are hard!

## On the shoulders of giants

Over the past decades the field of computing has largely settled on a few common
standards.

### Network Protocols

The vast majority of services work over the HTTP or HTTP/2 network protocols.
This solves problems such as:

* [URIs]
* [Sessions]
* [Status Codes] (eg. 404 Not Found)
* [Methods] (eg. GET, POST, PUT)
* [Headers] (eg. Authorization, User-Agent)
* [Pipelining]

TiKV uses **HTTP/2**. HTTP/2 is more performant and capable than HTT/1 for TiKVs uses.

### Interfacing

> With those abilities supported there remains a need to work with structures of

Reviewer 2: ```suggestion
With those abilities supported, there remains a need to work with structures of
```

data. Commonly this ends up being an *interface description language* format
like [Protocol Buffers] (protobufs), which TiKV uses. Unlike an *interchange*
*format*, an *interface description language* allows for the definition of
services and RPCs in addition to just data interchange. This solves the
following problems:

* [Encoding] (ASCII? UTF-8? UTF-16? WTF-8?)
* Serialization/Deserialization format (Text-to-`struct` and vice versa)
* Backward/Forward compatibility (Eg Field data type changed)
* Service & RPC definition

### Wrapping it all together

> Simply having the pieces is not enough, making it usable for all parties

Reviewer 2: ```suggestion
Simply having the pieces is not enough. Making it usable for all parties
```
